### PR TITLE
Resolve #715 Metric for unknown probe

### LINKED ISSUE #715: [Feature Request] > Metric for unknown module
Currently this is hidden in the response
https://github.com/prometheus/blackbox_exporter/blob/9fd70341277528877d55bfeffd46593646bca347/main.go#L77

Would be great to add two info points to this
- Log Error message
- Metric

Reason we ask for a metric is so that we(SRE) team can alert on misconfiguration of the module by teams. Thanks

I might get time to do this myself...seems easy enough 🤞

### PR DESCRIPTION
**Pull request by gavk34**
Resolve #715

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,6 +16,7 @@ _test
 _cgo_defun.c
 _cgo_gotypes.go
 _cgo_export.*
+.idea
```
Comment (brian-brazil) [inline .gitignore:19]: This belongs in your global gitignore.

File: README.md [hunk 2]
```diff
@@ -31,6 +31,8 @@ will return metrics for a HTTP probe against google.com. The `probe_success`
 metric indicates if the probe succeeded. Adding a `debug=true` parameter
 will return debug information for that probe.
 
+Note the `probe_module_unknown` metric indicates if the module used is unknown.
```
Comment (brian-brazil) [inline README.md:34]: This doesn't belong here, please remove it.

File: main.go [hunk 3]
```diff
@@ -68,13 +68,22 @@ var (
 )
 
 func probeHandler(w http.ResponseWriter, r *http.Request, c *config.Config, logger log.Logger, rh *resultHistory) {
+	probeModuleUnknownGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_module_unknown",
+		Help: "Displays whether or not probe module is unknown",
+	})
+	registry := prometheus.NewRegistry()
```
Comment (brian-brazil) [inline main.go:75]: This isn't right. This could be a counter using the default registry, and defined at file level.

File: main.go [hunk 4]
```diff
@@ -68,13 +68,22 @@ var (
...
+	registry.MustRegister(probeModuleUnknownGauge)
+
 	moduleName := r.URL.Query().Get("module")
 	if moduleName == "" {
 		moduleName = "http_2xx"
 	}
 	module, ok := c.Modules[moduleName]
 	if !ok {
 		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), http.StatusBadRequest)
+		level.Error(logger).Log("msg", "Error ", "err", fmt.Sprintf("unknown module %s", moduleName))
```
Comment (brian-brazil) [inline main.go:85]: No need to spam the logs when someone has incorrect usage.

Comment (gavk34) [reply main.go:85]: Remove or switch to `debug`?

File: .gitignore (hunk 1 above)
Comment (gavk34) [reply .gitignore]: Removed 👍

File: README.md (hunk 2 above)
Comment (gavk34) [reply README.md]: Done 👍

File: main.go (hunk 3 above)
Comment (gavk34) [reply main.go]: Think this is done now, please review

File: main.go (hunk 4 above)
Comment (gavk34) [reply main.go]: I have switched to `debug`

File: main.go [hunk 5]
```diff
@@ -65,6 +65,11 @@ var (
 		"icmp": prober.ProbeICMP,
 		"dns":  prober.ProbeDNS,
 	}
+
+	probeModuleUnknownCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "probe_module_unknown",
```
Comment (brian-brazil) [inline main.go:70]: Counters should end with _total.

This is not a metric from a probe, it is a metric about the blackbox exporter. Thus it should be blackbox_

File: main.go [hunk 6]
```diff
@@ -75,6 +80,8 @@ func probeHandler(w http.ResponseWriter, r *http.Request, c *config.Config, logg
 	module, ok := c.Modules[moduleName]
 	if !ok {
 		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), http.StatusBadRequest)
+		level.Debug(logger).Log("msg", "Error ", "err", fmt.Sprintf("unknown module %s", moduleName))
```
Comment (brian-brazil) [inline main.go:83]: The message should be a fixed string about what's going on, and then the module a pair on its own.

File: main.go [hunk 7]
```diff
@@ -80,7 +80,7 @@ func probeHandler(w http.ResponseWriter, r *http.Request, c *config.Config, logg
 	module, ok := c.Modules[moduleName]
 	if !ok {
 		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), http.StatusBadRequest)
-		level.Debug(logger).Log("msg", "Error ", "err", "unknown module name", "module", moduleName)
+		level.Debug(logger).Log("msg", "Error finding module", "module", moduleName)
```
Comment (brian-brazil) [inline main.go:83]: It's not about finding, keep it consistent with the error we send back over HTTP - so Unknown module.

Comment (gavk34) [reply main.go]: Done, just in time for the Toy Show 🇮🇪 😄

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**